packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,181 +474,663 @@ class CommonFinders {
+class FinderResult<CandidateType> extends Iterable<CandidateType> {
+  /// Creates a new [FinderResult] that describes the `values` using the given
+  /// `describeMatch` callback.
+  ///
+  /// {@macro flutter_test.finders.FinderBase.describeMatch}
+  FinderResult(DescribeMatchCallback describeMatch, Iterable<CandidateType> values)
+    : _describeMatch = describeMatch, _values = values;
+
+  final DescribeMatchCallback _describeMatch;
+  final Iterable<CandidateType> _values;
+
+  @override
+  Iterator<CandidateType> get iterator => _values.iterator;
+
+  @override
+  String toString() {
+    final List<CandidateType> valuesList = _values.toList();
+    // This will put each value on its own line with a comma and indentation
+    final String valuesString = valuesList.fold(
+      '',
+      (String current, CandidateType candidate) => '$current\n  $candidate,',
     );
+    return 'Found ${valuesList.length} ${_describeMatch(Plurality._fromNum(valuesList.length))}: ['
+      '${valuesString.isNotEmpty ? '$valuesString\n' : ''}'
+      ']';
   }
+}
+
+/// Provides backwards compatibility with the original [Finder] API.
+mixin _LegacyFinderMixin on FinderBase<Element> {
+  Iterable<Element>? _precacheResults;
 
-  Iterable<Element>? _cachedResult;
+  /// Describes what the finder is looking for. The description should be
+  /// a brief English noun phrase describing the finder's requirements.
+  @Deprecated(
+    'Use FinderBase.describeMatch instead. '
+    'FinderBase.describeMatch allows for more readable descriptions and removes ambiguity about pluralization. '
+    'This feature was deprecated after v3.13.0-0.2.pre.'
+  )
+  String get description;
 
-  /// Returns the current result. If [precache] was called and returned true, this will
-  /// cheaply return the result that was computed then. Otherwise, it creates a new
-  /// iterable to compute the answer.
+  /// Returns all the elements in the given list that match this
+  /// finder's pattern.
   ///
-  /// Calling this clears the cache from [precache].
-  Iterable<Element> evaluate() {
-    final Iterable<Element> result = _cachedResult ?? apply(allCandidates);
-    _cachedResult = null;
-    return result;
+  /// When implementing Finders that inherit directly from
+  /// [Finder], [findInCandidates] is the main method to override. This method
+  /// is maintained for backwards compatibility and will be removed in a future
+  /// version of Flutter. If the finder can efficiently be described just in
+  /// terms of a predicate function, consider mixing in [MatchFinderMixin]
+  /// instead.
+  @Deprecated(
+    'Override FinderBase.findInCandidates instead. '
+    'Using the FinderBase API allows for more consistent caching behavior and cleaner options for interacting with the widget tree. '
+    'This feature was deprecated after v3.13.0-0.2.pre.'
+  )
+  Iterable<Element> apply(Iterable<Element> candidates) {
+    return findInCandidates(candidates);
   }
```

Comment:
@goderbauer

To prevent breaking changes, I ended up implementing _both_ `apply` and `findInCandidates` here. They're both implemented to reference the other since they're a direct replacement. Existing `Finder`s must already implement `apply`, resulting in `findInCandidates` using that implementation. When new `Finder`s are implemented, they are expected to override `findInCandidates`, which will result in `apply` using that implementation. Once everyone has migrated to using `findInCandidates` and `apply` has been removed, `findInCandidates` will be abstract as expected for `FinderBase`.

The issue I have is that this makes neither `apply` nor `findInCandidates` actually abstract for the deprecation period. This means that new `Finder` implementations _could_ implement neither. Do you have any suggestions for a better way to go forward here?

Reply:
It should be fine to just document this explicitly.

Reply:
Or we could have a debug flag that `apply` and `findInCandidates` set to true before calling the other methods. At the beginning of both methods we check that the flag is false, if it is true, it means someone forgot to override either `apply` or `findInCandidates`. At the end of both methods, we need to reset the flag back to false again.

But honestly, I think proper docs are fine here.

Reply:
The class documentation points users to override `findInCandidates`, so resolving this.